src/stream/src/executor/upstream_sink_union.rs

```diff
@@ -311,25 +422,79 @@ impl UpstreamSinkUnionExecutor {
                 "UpstreamSinkUnion",
             ]);
 
-        let upstreams = DynamicReceivers::new(inputs, Some(barrier_align), None);
+        let upstreams = DynamicReceivers::new(inputs, Some(barrier_align.clone()), None);
         pin_mut!(upstreams);
 
-        while let Some(msg) = upstreams.next().await {
-            yield msg?;
+        let mut current_barrier = None;
+
+        // Here, `tokio::select` cannot be used directly in `try_stream` function.
+        // Err for breaking the loop. Ok(None) for continuing the loop.
```

Comment:
Can we have a loop inside the async closure so that it can return a cleaner `Result<Message>` instead of `Ok(None)`?